### PR TITLE
Dismiss button for banner

### LINKED ISSUE #428: Add close button for the recruitment banner.
We have a recruitment banner rendered in CourseRenderPane. We want there to be a button to close that banner, and it should save the user's decision so that if they close their browser window and come back, it should stay hidden if they've pressed close before. You can do this with localStorage. If you look for `window.localStorage` in the project you can find some examples.
screenshot with the desired design crudely drawn on
It should look something like this, but the design is flexible so you can do whatever you want as long as it makes sense. Make sure to post a screenshot when you open a PR.

### PR DESCRIPTION
**Pull request by MinhxNguyen7**
## Summary
Added button to dismiss recruitment banner. Dismissal persists in `window.localStorage.dismissedRecruitmentBanner`

image

## Test Plan
Check that the banner disappears after click and does not re-display in subsequent sessions.

## Issues
Closes #428

### CODE REVIEW

**Comment by Voark** [comment]
Whoops I need to fix this action, forgot it doesn't work when merging from branches outside this repo. Anyways my only comment is could we add an outline so it looks more like a banner? Otherwise your code and changes look great!

**Comment by MinhxNguyen7** [comment]
Thanks! Here it is (7221802):
image

File: src/components/RightPane/CoursePane/CourseRenderPane.js [hunk 1]
```diff
@@ -89,17 +91,44 @@ const flattenSOCObject = (SOCObject) => {
...
-    </div>
-);
+const RecruitmentBanner = ({ className }) => {
+    // Idk how else to force a function component to update
+    const [, updateState] = React.useState();
+    const forceUpdate = React.useCallback(() => updateState({}), []);
+
+    let displayRecruitmentBanner =
+        !Boolean(window.localStorage.getItem('dismissedRecruitmentBanner')) &&
+        ['COMPSCI', 'IN4MATX', 'I&C SCI', 'STATS'].includes(RightPaneStore.getFormData().deptValue);
```
Comment (ChaseC99) [inline src/components/RightPane/CoursePane/CourseRenderPane.js]: I think this is how you would do it without needing a useCallback function
```suggestion
    const [displayRecruitmentBanner, setDisplayRecruitmentBanner] = useState(
        !Boolean(window.localStorage.getItem('dismissedRecruitmentBanner')) &&
        ['COMPSCI', 'IN4MATX', 'I&C SCI', 'STATS'].includes(RightPaneStore.getFormData().deptValue);
    )
```

File: src/components/RightPane/CoursePane/CourseRenderPane.js [hunk 2]
```diff
@@ -89,17 +91,44 @@ const flattenSOCObject = (SOCObject) => {
...
+                            Join ICSSC and work on AntAlmanac and other projects!
+                        </a>
+                        <br />
+                        We have positions for experienced devs and those with zero experience!
+                    </div>
+
+                    <Button
+                        onClick={() => {
+                            window.localStorage.setItem('dismissedRecruitmentBanner', true);
+                            forceUpdate();
```
Comment (ChaseC99) [inline src/components/RightPane/CoursePane/CourseRenderPane.js]: Then down here we can just set the display state to false
```suggestion
                            setDisplayRecruitmentBanner(false);
```

File: src/components/RightPane/CoursePane/CourseRenderPane.js [hunk 3]
```diff
@@ -228,10 +257,8 @@ class CourseRenderPane extends PureComponent {
 
             currentView = (
                 <div className={classes.root}>
-                    <div className={classes.bannerContainer}>
-                        {['COMPSCI', 'IN4MATX', 'I&C SCI', 'STATS'].includes(
-                            RightPaneStore.getFormData().deptValue
-                        ) && <RecruitmentBanner className={classes.banner} />}
+                    <div className={classes.bannerContainer} style={{ paddingRight: '0', marginBottom: '5px' }}>
```
Comment (ChaseC99) [inline src/components/RightPane/CoursePane/CourseRenderPane.js]: These styles should be added into the `bannerContainer` class

File: src/components/RightPane/CoursePane/CourseRenderPane.js [hunk 4]
```diff
@@ -89,17 +91,44 @@ const flattenSOCObject = (SOCObject) => {
...
+const RecruitmentBanner = ({ className }) => {
+    // Idk how else to force a function component to update
+    const [, updateState] = React.useState();
+    const forceUpdate = React.useCallback(() => updateState({}), []);
+
+    let displayRecruitmentBanner =
+        !Boolean(window.localStorage.getItem('dismissedRecruitmentBanner')) &&
+        ['COMPSCI', 'IN4MATX', 'I&C SCI', 'STATS'].includes(RightPaneStore.getFormData().deptValue);
+
+    const bannerStyle = { flexDirection: 'left', justifyContent: 'flex-end' };
```
Comment (ChaseC99) [inline src/components/RightPane/CoursePane/CourseRenderPane.js]: This should be moved to the `styles` object at the top of the file

**Review by ChaseC99 — CHANGES_REQUESTED** [review]
Looking solid!

One thing I'm wondering - if they click the `x`, they'll never see the recruitment banner again.   Like ever. Is this what we want? I wonder if there should be a timeout (a quarter / year) before it's shown again, or if `dismissedRecruitmentBanner` should just be saved to AppStore so that it resets on page refresh.

**Comment by EricPedley** [comment]
> One thing I'm wondering - if they click the `x`, they'll never see the recruitment banner again. Like ever. Is this what we want? I wonder if there should be a timeout (a quarter / year) before it's shown again, or if `dismissedRecruitmentBanner` should just be saved to AppStore so that it resets on page refresh.

Good point. One quarter timeout sounds like the best option to me. As for implementation, maybe add a timestamp to the local storage object and use that to check if the state is expired?

File: src/components/RightPane/CoursePane/CourseRenderPane.js [hunk 5]
```diff
@@ -89,17 +91,44 @@ const flattenSOCObject = (SOCObject) => {
...
+
+    return displayRecruitmentBanner ? (
+        <>
+            <Paper elevation={1} square>
```
Comment (EricPedley) [inline src/components/RightPane/CoursePane/CourseRenderPane.js]: Didn't know about this `elevation` thing for box shadow. Really neat! (you can just resolve this comment right away)

File: src/components/RightPane/CoursePane/CourseRenderPane.js (hunk 3 above)
Comment (EricPedley) [inline src/components/RightPane/CoursePane/CourseRenderPane.js]: Now that the conditional rendering is inside the component, we can just put this wrapper div in the component as well.

**Review by EricPedley — CHANGES_REQUESTED** [review]
Address the comments and make the dismissal only persist for a quarter, and this PR will be good to go.
Also, nice job getting to this so fast. I was slightly worried about how to add the "Tired of this banner, write the close button yourself!" line to the banner without running out of space but it's a non-issue now.

**Comment by MinhxNguyen7** [comment]
> Address the comments and make the dismissal only persist for a quarter

How should I define the quarter? 11 weeks?

**Comment by EricPedley** [comment]
> How should I define the quarter? 11 weeks?

Good question. 11 weeks sounds fine.

**Comment by EricPedley** [comment]
Now that the `recruitment-banner` PR is merged, I changed the base branch of this PR to `main`. You now need to rebase your commits on the fork on top of `origin/main` and force-push to your branch to rewrite the history to include the squash-merged commit on `main`. Here is a diagram of what I'm talking about (not to scale w.r.t. number of commits):
git history diagram
This tutorial explains rebasing in basic detail: https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase. Let us know if you need help with this. I could explain it in more excruciating detail. Also, I say this whole thing with 90% but not 100% confidence. If we don't need to rebase here, someone please comment.

**Comment by MinhxNguyen7** [comment]
Admittedly, I'm not sure how to rebase. I figured I could just pull from `main` and resolve the conflict (92c0c) so that by branch can be merged intto `main`. Please do tell me if I messed something up. Thanks!

File: src/components/AppBar/ImportStudyList.js [hunk 6]
```diff
@@ -133,9 +134,11 @@ class ImportStudyList extends PureComponent {
         return (
             <>
                 {/* TODO after mui v5 migration: change icon to ContentPasteGo */}
-                <Button onClick={this.handleOpen} color="inherit" startIcon={<PostAdd />}>
-                    Import
-                </Button>
+                <Tooltip title="Import a schedule from your Study List">
```
Comment (EricPedley) [inline src/components/AppBar/ImportStudyList.js]: Looks like your tooltip change snuck in here. I'm gonna be nitpicky here and ask that you undo the change here and raise a separate PR for it. I want to enforce [atomic commits](https://www.aleksandrhovhannisyan.com/blog/atomic-git-commits/) for the repo, and although this is a small change, I want to avoid the [broken window effect](https://storck.io/posts/broken-windows-theory-in-software/)

File: src/components/RightPane/CoursePane/CourseRenderPane.js [hunk 7]
```diff
@@ -91,43 +96,45 @@ const flattenSOCObject = (SOCObject) => {
...
-const RecruitmentBanner = ({ className }) => {
+const RecruitmentBanner = (classes) => {
     // Idk how else to force a function component to update
     const [, updateState] = React.useState();
     const forceUpdate = React.useCallback(() => updateState({}), []);
 
+    // Display recruitment banner if more than 11 weeks (in ms) has passed since last dismissal
     let displayRecruitmentBanner =
-        !Boolean(window.localStorage.getItem('dismissedRecruitmentBanner')) &&
+        Date.now() - window.localStorage.getItem('recruitmentDismissalTime') > 6652800000 &&
```
Comment (EricPedley) [inline src/components/RightPane/CoursePane/CourseRenderPane.js]: nit: the dismissal time would be easier to edit as an expression instead of the calculated result. e.g. `11 * 7 * 24 * 3600 * 1000`. This is kinda opinionated though, so you can change this if you agree or leave it as-is.

**Review by EricPedley — CHANGES_REQUESTED** [review]
Split the tooltip changes into a separate PR and respond to the nitpick about the dismissal time, then this'll be good to go. The messy git history is fine since we use squash-and-merge for merging PRs, so no need to do the rebase. That being said, I'll make a little wiki post demonstrating how to do it anyways since it's a bit complicated in this case, but it's a good skill to have and this will provide a really good real-world example for others to reference. Clean git history makes a PR much easier to review IMO so I would like as many contributors as possible to be comfortable with rebasing to clean up their history.

File: src/components/RightPane/CoursePane/CourseRenderPane.js (hunk 7 above)
Comment (ChaseC99) [reply src/components/RightPane/CoursePane/CourseRenderPane.js]: ^ This would make it easy to change the number of weeks later on

File: src/components/AppBar/ImportStudyList.js (hunk 6 above)
Comment (MinhxNguyen7) [reply src/components/AppBar/ImportStudyList.js]: Sorry about that. I've reverted it, which should be ok since you're squashing anyway right?

**Review by EricPedley — APPROVED** [review]

**Comment by EricPedley** [comment]
Looks like all Chase's concerns were met so I'm gonna override and merge.
